### PR TITLE
2331: Improve review notes

### PR DESCRIPTION
**Pull request by pavelrappo**
There are two variants of a note that can be added to an item in the "Reviewers" list of a PR body:

  - 🔄 Re-review required ...
  - ⚠️ Review applies to ...

That note may confuse or draw unneeded attention to the PR body.

Firstly, as a result of recently integrated [SKARA-2312][], "⚠️ Review applies to ..." is also added to an item (i.e. review) that applies to the head of the PR. Secondly, "🔄 Re-review required (review applies to ..." may be present for some items even though a PR has the green "ready" label.

[SKARA-2312]: https://bugs.openjdk.org/browse/SKARA-2312

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [SKARA-2331](https://bugs.openjdk.org/browse/SKARA-2331): Improve review notes (**Bug** - P4)

### Reviewers
 * [Zhao Song](https://openjdk.org/census#zsong) (@zhaosongzs - **Reviewer**) ⚠️ Review applies to [cd128621](https://git.openjdk.org/skara/pull/1679/files/cd1286215f146eca53dda62da07ab2ec65a3e0fd)
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to [f4dde9b4](https://git.openjdk.org/skara/pull/1679/files/f4dde9b421993259f1c86b79314c1a494a3a1b3d)
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**) ⚠️ Review applies to [f4dde9b4](https://git.openjdk.org/skara/pull/1679/files/f4dde9b421993259f1c86b79314c1a494a3a1b3d)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/skara.git pull/1679/head:pull/1679` \
`$ git checkout pull/1679`

Update a local copy of the PR: \
`$ git checkout pull/1679` \
`$ git pull https://git.openjdk.org/skara.git pull/1679/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1679`

View PR using the GUI difftool: \
`$ git pr show -t 1679`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/skara/pull/1679.diff">https://git.openjdk.org/skara/pull/1679.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/skara/pull/1679#issuecomment-2236214613)

### CODE REVIEW

**Comment by liach** [comment]
Can we change the emoji for "outdated review" from :warning: to something weaker than :arrows_counterclockwise:, like :clock3:?

**Comment by pavelrappo** [comment]
> Can we change the emoji for "outdated review" from ⚠️ to something weaker than 🔄, like 🕒?

The "⚠️" icon indicates that a particular review is stale: that is, the review applies to a commit which is no longer the HEAD of the PR. Although it draws attention, this indication does not necessitate any action. Such a review still counts: it is legitimate in every way.

The "🔄" icon also indicates that a particular review is stale. However, unlike the previous type, such a review is restricted. While the reviewer will be mentioned in the "Reviewed-by" list of the commit message, currently the review is not counted towards the minimum number of reviews and, thus, does not help to get the PR closer to "ready".

In my opinion, the only circumstance where "🔄" seems too "strong" of an indication is when a PR is "ready". This happens when there are more reviews than required, but some of the extra ones are stale.

This PR addresses that issue. Does it also address your issue?

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java [hunk 1]
```diff
@@ -1343,7 +1347,8 @@ private void checkStatus() {
             } else {
                 // Determine current status
                 jcheckType = "target jcheck";
-                checkablePullRequest.executeChecks(localHash, censusInstance, visitor, targetJCheckConf);
+                var issues = checkablePullRequest.executeChecks(localHash, censusInstance, visitor, targetJCheckConf);
+                tooFewReviewers = issues.stream().anyMatch(TooFewReviewersIssue.class::isInstance);
```
Comment (zhaosongzs) [inline bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:1351]: I am not sure if it's a good idea to change the return value of `CheckablePullRequest.executeChecks`.

I think you can use `visitor.getChecks()` to determine if reviewers check failed or not.

```
        var tooFewReviewers = visitor.getChecks()
                .entrySet().stream()
                .anyMatch(entry -> entry.getKey().contains("Change must be properly reviewed") && entry.getValue().equals(false));
```

Comment (pavelrappo) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:1351]: > I am not sure if it's a good idea to change the return value of `CheckablePullRequest.executeChecks`.

Why?

> I think you can use `visitor.getChecks()` to determine if reviewers check failed or not.

Keys from `Map<String, Boolean> getChecks()` are to be displayed to an end user; they don't seem to be for programmatic consumption. I shudder when thinking about parsing these:

    "Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))"
    "Change must not contain extraneous whitespace"

Comment (zhaosongzs) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:1351]: I was just thinking we should make the minimal necessary changes.  I admit that parsing text displayed to the end user isn't best practice, but we do it in many places of SKARA.

Comment (zhaosongzs) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:1351]: It's just a suggestion, since you don't like it and I am ok with your pr, I will approve it now.

**Review by zhaosongzs — APPROVED** [review]

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java (hunk 1 above)
Comment (pavelrappo) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:1351]: Usually, making a minimally necessary change is good, and parsing end-user text is bad. In fact, I'd go to great lengths to avoid "stringly typed" code that operates on strings that are also subject to change.

Here, I initially thought I might be violating some encapsulation boundary by bringing a dependency. Maybe I still am. If nothing else, it's not the first class imported from that package:

     import org.openjdk.skara.jcheck.JCheckConfiguration;
    +import org.openjdk.skara.jcheck.TooFewReviewersIssue;

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java [hunk 2]
```diff
@@ -607,12 +608,14 @@ private Optional<String> getReviewersList() {
...
-                                                       + "](" + pr.filesUrl(hash.get()) + "))";
-                                           } else {
-                                               entry += " ⚠️ Review applies to [" + hash.get().abbreviate()
-                                                       + "](" + pr.filesUrl(hash.get()) + ")";
+                                           if (!hash.get().equals(pr.headHash())) {
+                                               if (!reviewCoverage.covers(review) && tooFewReviewers) {
+                                                   entry += " 🔄 Re-review required (review applies to [" + hash.get().abbreviate()
+                                                           + "](" + pr.filesUrl(hash.get()) + "))";
+                                               } else {
+                                                   entry += " ⚠️ Review applies to [" + hash.get().abbreviate()
```
[the remark this answers is not in the document]
Comment (pavelrappo) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:616]: Rest assured, I will not integrate this PR until we figure out what to do with the issue you raised, Chen. That issue should be fixed either in this PR or a follow-up one.

Personally, I think if a review is stale but otherwise good, we don't need to add any emoji. It suffices to add the commit hash the review applies to, and maybe indicate how far away that commit is from HEAD, to quantify staleness: for example, "3 commits behind". Any extra commits added since that review may only be simple merges anyway.

However, if a review is stale and restricted (i.e. does not count in required reviews), we may add "⚠️". In a sense, it would provide a similar indication to that of a repo configured with:

    useStaleReviews=true
    acceptSimpleMerges=false

In this picture, the "🔄" emoji goes away completely.

[the remark this answers is not in the document]
Comment (pavelrappo) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:616]: To add to my previous message.

We should keep in mind that any indication in question is primarily for a PR _author_. From my own experience as a _reviewer_, I don't normally follow a particular PR that I have reviewed to see if my review has become stale and needs an update. If the author feels I should re-review their PR, they ping me directly.

[the remark this answers is not in the document]
Comment (kevinrushforth) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:616]: I was thinking along similar lines. Basically there are two main cases, `useStaleReviews=(true|false)`. Here is logic that I think does roughly what we want:

```
if (useStaleReviews) {
    display the "⚠️" emoji and "Review applies to HASH" message always for stale reviews
} else {
    display the "🔄" emoji and "Re-review required" message for all stale reviews
    if (acceptSimpleMerges) {
        display the "review applies to HASH" message without any emoji for non-HEAD simple merge reviews [1]
    }
}

[1] By this I mean a review of a commits that isn't the HEAD, but that differs
from the head only by a simple merge
```

This ensures that the same PR doesn't show both the "⚠️" and "🔄" emojis.

Another possible incremental improvement we could consider is to only display the re-review needed emoji + message if the review criteria isn't satisfied. This would mean that as soon the minimum number of Reviewers re-review, the rest of the stale reviews lose the emoji (but retain a message indicating that they are out of date).

```
if (useStaleReviews) {
    display the "⚠️" emoji and "Review applies to HASH" message always for stale reviews
} else {
    if (review criteria not satisfied) {
        display the "🔄" emoji and "Re-review required" message for all stale reviews
        if (acceptSimpleMerges) {
            display the "review applies to HASH" message without an emoji for non-HEAD simple merge reviews [1]
        }
    } else {
        display the "review applies to HASH" message without an emoji for stale reviews and non-HEAD simple merge reviews [1]
    }
}
```

[the remark this answers is not in the document]
Comment (pavelrappo) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:616]: > [1] By this I mean a review of a commits that isn't the HEAD, but that differs
from the head only by a simple merge

I would clarify that trailing bit as follows:

> but that differs from the head only by commits that are simple merges

This is to convey that for our purposes a series of adjacent simple merges is no different from a single simple merge.

Back to your proposal. I'll try to implement the second approach because it's in line with what this PR already does.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java [hunk 3]
```diff
@@ -596,31 +597,43 @@ private String warningListToText(List<String> additionalErrors) {
...
-                                   return entry;
-                               })
-                               .collect(Collectors.joining("\n"));
+                .filter(review -> review.verdict() == Review.Verdict.APPROVED)
+                .map(review -> {
+                    var entry = " * " + formatReviewer(review.reviewer());
+                    if (!review.targetRef().equals(pr.targetRef())) {
+                        if (useStaleReviews || tooFewReviewers) {
+                            entry += " 🔄 Re-review required (review was made when pull request targeted the [" + review.targetRef()
+                                    + "](" + pr.repository().webUrl(new Branch(review.targetRef())) + ") branch)";
```
Comment (zhaosongzs) [inline bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:608]: Why should we always display  "🔄" when `useStaleReviews==true`?

Comment (zhaosongzs) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:608]: I see, previously we always require re-review when target branch changes.
My question should be why don't we display "🔄" when `useStaleReviews==false`?

**Review by zhaosongzs — APPROVED** [review]

**Comment by pavelrappo** [comment]
If no concerns are raised until 18:00 UTC, the 24th of July, I'll integrate this PR. Thanks.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java [hunk 4]
```diff
@@ -596,31 +597,43 @@ private String warningListToText(List<String> additionalErrors) {
...
+                        } else {
+                            entry += " Review was made when pull request targeted the [" + review.targetRef()
+                                    + "](" + pr.repository().webUrl(new Branch(review.targetRef())) + ") branch";
+                        }
+                    } else {
+                        var hash = review.hash();
+                        if (hash.isPresent()) {
+                            if (!hash.get().equals(pr.headHash())) {
+                                if (reviewCoverage.covers(review) || !tooFewReviewers) {
+                                    entry += (tooFewReviewers ? " ⚠️ " : " ")
```
[the remark this answers is not in the document]
Comment (pavelrappo) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:618]: `useStaleReviews` is baked into `reviewCoverage.covers(review)`:

    public boolean covers(Review review) {
        return cachedCoverage.computeIfAbsent(review, this::covers0);
    }

    private boolean covers0(Review review) {
        var r = review.hash();
        // Reviews without a hash are never valid as they referred to no longer
        // existing commits.
        if (r.isEmpty() || review.verdict() != Review.Verdict.APPROVED
                || !review.targetRef().equals(pr.targetRef())) {
            return false;
        }
        if (useStaleReviews || r.get().equals(pr.headHash())) {
            return true;
        }
        if (!acceptSimpleMerges) {
            return false;
        }

Here and elsewhere, `tooFewReviewers` is checked because it's decisive: if there are enough reviewers, no emoji or the word "required" should be used.

[the remark this answers is not in the document]
Comment (kevinrushforth) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:618]: > If there are enough reviewers, no emoji or the word "required" should be used.

I think I now understand what you are trying to do, although I don 't think I agree with the premise. If I understand the logic (which I think I do now), here are two cases that won't do what I would expect:

1. `useStaleReviews=false` : if there are too few reviewers, simple merges will show a warning icon
2. `useStaleReviews=true` : if there are enough reviewers, all other reviews will not get the warning icon

I think 2 basically makes the warning icon useless (and 1 is a bug)

[the remark this answers is not in the document]
Comment (pavelrappo) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:618]: Let's dissect it. When you are saying you don't agree with the premise, do you mean you don't agree with this?

> If there are enough reviewers, no emoji or the word "required" should be used.

If so, please clarify. How do you want it to be?

[the remark this answers is not in the document]
Comment (pavelrappo) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:618]: @kevinrushforth ping

[the remark this answers is not in the document]
Comment (kevinrushforth) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:618]: Sorry for the delay.

For the (now largely unused) case of `useStaleReviews=true`, I would expect no change from today's behavior: Regardless of whether or not the review criteria is met, print "⚠️ Review applies to HASH" for all stale reviews.

For the (now widely-used) case of `useStaleReviews=false`: If the review criteria is _not_ met, print "🔄 Re-review required (review applies to HASH)" for all stale reviews; if the review criteria _is_ met, print "Review applies to HASH" with no emoji for all stale reviews.

[the remark this answers is not in the document]
Comment (pavelrappo) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:618]: To further clarify, the algorithm you propose applies to the usual case where some commits have been added on top of those reviewed; and not to the other two cases, where either the reviewed commit is no longer present or the PR has changed its target branch. Is that correct?

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java (hunk 1 above)
Comment (erikj79) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:1361]: I think it's ok to return the list of jcheck issues, but the variable name `issues` can be confusing as an `Issue` in the bots is usually some form of Jira Issue. I would recommend `jcheckIssues` to avoid confusion.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (erikj79) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:618]: I'm ok with Kevin's proposed behavior, but I'm not sure what the significance of your last question is, Pavel. Are we behaving differently today when the commit has gone missing or the branch has changed?

[the remark this answers is not in the document]
Comment (erikj79) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:618]: Also, when this is all done, could you summarize the complete behavior in the bug description, or comment, similar to how Kevin presented it here?

[the remark this answers is not in the document]
Comment (pavelrappo) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:618]: > I'm not sure what the significance of your last question is, Pavel. Are we behaving differently today when the commit has gone missing or the branch has changed?

In either of those two cases, in master, we will display the "🔄" emoji and the word "required" regardless of whether the PR has been sufficiently reviewed:

https://github.com/openjdk/skara/blob/45468cde2dc5273c70f728c46c6b133e3d16b55a/bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java#L602-L622

If we are trying to generally improve review notes, we perhaps should treat those two cases similarly to our main/typical case?

[the remark this answers is not in the document]
Comment (erikj79) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:618]: I agree, those cases should be folded into the new behavior with regards to emojis.

**Comment by pavelrappo** [comment]
I hope I've addressed everyone's concerns. @kevinrushforth, @erikj79, @zhaosongzs, please re-review.

**Review by zhaosongzs — APPROVED** [review]
Looks good!
/reviewers 2

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java (hunk 1 above)
Comment (pavelrappo) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:1363]: Fair enough. I can see that "issue" is an overloaded term in the Skara domain:

  - org.openjdk.skara.issuetracker.Issue
  - org.openjdk.skara.jcheck.Issue

One word, two meanings. Does `jcheckIssues` sound better to you?

Comment (erikj79) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:1363]: Yes, I wrote just that in the previous comment from last week. :)

Comment (pavelrappo) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:1363]: > Yes, I wrote just that in the previous comment from last week. :)

D'oh. The PR has been open for quite a while, and I forgot that you asked that question. While I cannot guarantee that I didn't internalise and then forgot your question, it's good that we're now on the same page.

**Review by erikj79 — APPROVED** [review]

**Review by kevinrushforth — APPROVED** [review]
Looks good now.
